blueocean-pipeline-api-impl/src/test/java/io/jenkins/blueocean/rest/impl/pipeline/PipelineNodeTest.java

```diff
@@ -2462,6 +2402,78 @@ public void parallelStagesGroupsAndNestedStages() throws Exception {
+    @Test
+    @Issue("JENKINS-53816")
+    public void graphConsistentWhileExecuting() throws Exception {
+
+        final String expectedNodeNames =
+            "A, B, B-A, B-B, B-B-1, B-B-2, B-C, B-C-1, B-C-2, C, D, D-A, D-B, D-B-1, D-B-2, D-C, D-C-1, D-C-2";
+
+        String completeNodeNames = checkConsistencyWhileBuilding("JENKINS-53816.jenkinsfile");
+
+        assertEquals("node names", expectedNodeNames, completeNodeNames);
+    }
+
+    @Test
+    @Issue("JENKINS-53816")
+    public void graphConsistentWhileExecuting2() throws Exception {
+
+        final String expectedNodeNames = "first-sequential-stage, first-solo, multiple-stages, other-single-stage, " +
+            "parent, second-sequential-stage, second-solo, single-stage, third-sequential-stage";
+
+        String completeNodeNames = checkConsistencyWhileBuilding("sequential_parallel_stages_long_run_time.jenkinsfile");
+
+        assertEquals("node names", expectedNodeNames, completeNodeNames);
+    }
+
+    private String checkConsistencyWhileBuilding(String jenkinsFileName) throws Exception {
+
+        /*
+            Run a complex pipeline to completion, then start a new build and inspect it while it's running, to exercise
+            the code that merges incomplete runs with previous builds to generate a complete graph
+         */
+
+        WorkflowJob p = createWorkflowJobWithJenkinsfile(getClass(), jenkinsFileName);
+
+        // Do an initial run, collect the nodes
+        final WorkflowRun run1 = p.scheduleBuild2(0).waitForStart();
+        j.waitForCompletion(run1);
+
+        final List<BluePipelineNode> completeNodes = new PipelineNodeContainerImpl(run1, new Link("foo")).getNodes();
+
+        final String completeNodeNames = completeNodes.stream()
+                                                      .map(BluePipelineStep::getDisplayName)
+                                                      .sorted()
+                                                      .collect(Collectors.joining(", "));
+
+        // Start another build...
+        final WorkflowRun run2 = p.scheduleBuild2(0).waitForStart();
+
+        // ... then watch while it runs, checking for the same graph nodes
+
+        int loopCount = 0;
+
+        do {
+            Thread.sleep(500);
+
+            List<BluePipelineNode> runningNodes = new PipelineNodeContainerImpl(run2, new Link("foo")).getNodes();
+            String runningNodeNames = runningNodes.stream()
+                                                  .map(BluePipelineStep::getDisplayName)
+                                                  .sorted()
+                                                  .collect(Collectors.joining(", "));
+
+            assertEquals("running node names", completeNodeNames, runningNodeNames);
+
+            loopCount++;
+
+        } while (run2.isBuilding());
+
+        // Sanity check, make sure we're *actually* checking stuff.
+        assertTrue("Checked multiple times while building", loopCount > 5);
```

Comment:
I wonder if maybe we could have lower numbers on very slow/not powerful machines?

Reply:
I don't think it's a good idea. What use is a green test that's not really testing anything. I'd like to be able to use some other mechanism that time delays in the pipeline, something deterministic, but I don't know what that could really be. FWIW, the count running on my machine was > 90, so 5 seems pretty low.